lib/src/background_usecase.dart

```diff
@@ -0,0 +1,172 @@
+/// `data` parameter, while errors can be reported through the `error` parameter.
+///
+/// In addition, a `done` flag can be set to indicate that the isolate has completed its task.
+///
+/// An example would be a usecase that performs matrix multiplication.
+/// ```dart
+/// class MatMulUseCase extends BackgroundUseCase<List<List<double>>, MatMulUseCaseParams> {
+///  @override
+///  buildUseCaseTask() {
+///    return matmul;
+///  }
+///
+///  static void matmul(BackgroundUseCaseParams params) async {
+///    MatMulUseCaseParams matMulParams = params.params as MatMulUseCaseParams;
+///    List<List<double>> result = List<List<double>>.generate(
+///        10, (i) => List<double>.generate(10, (j) => 0));
+///
+///    for (int i = 0; i < matMulParams.mat1.length; i++) {
+///      for (int j = 0; j < matMulParams.mat1.length; j++) {
+///        for (int k = 0; k < matMulParams.mat1.length; k++) {
+///          result[i][j] += matMulParams.mat1[i][k] * matMulParams.mat2[k][j];
+///        }
+///      }
+///    }
+///    params.port.send(BackgroundUseCaseMessage(data: result));
+///
+///  }
+///}
+///
+/// ```
+/// Just like a regular [UseCase], a parameter class is recommended for any [BackgroundUseCase].
+/// An example corresponding to the above example would be
+/// ```dart
+/// class MatMulUseCaseParams {
+///  List<List<double>> mat1;
+///  List<List<double>> mat2;
+///  MatMulUseCaseParams(this.mat1, this.mat2);
+///  MatMulUseCaseParams.random() {
+///    var size = 10;
+///    mat1 = List<List<double>>.generate(size,
+///        (i) => List<double>.generate(size, (j) => i.toDouble() * size + j));
+///
+///    mat2 = List<List<double>>.generate(size,
+///        (i) => List<double>.generate(size, (j) => i.toDouble() * size + j));
+///  }
+///}
+/// ```
+abstract class BackgroundUseCase<T, Params> extends UseCase<T, Params> {
+  BehaviorSubject<T> _subject;
+  BackgroundUseCaseState _state = BackgroundUseCaseState.idle;
+  Isolate _isolate;
+  final ReceivePort _receivePort;
+  static UseCaseTask _run;
+
+  BackgroundUseCase()
+      : _receivePort = ReceivePort(),
+        _subject = BehaviorSubject(),
+        super() {
+    _receivePort.listen(_handleMessage);
+  }
+
+  BackgroundUseCaseState get state => _state;
+  bool get isRunning => _state != BackgroundUseCaseState.idle;
+
+  /// Executes the usecase on a different isolate. Spawns [_isolate]
+  /// using the static method provided by [buildUseCaseTask] and listens
+  /// to a [BehaviorSubject] using the [observer] provided by the user.
+  /// All [Params] are sent to the [_isolate] through [BackgroundUseCaseParams].
+  @override
+  void execute(Observer<T> observer, [Params params]) async {
+    if (!isRunning) {
+      _state = BackgroundUseCaseState.loading;
+      _subject.listen(observer.onNext,
+          onError: observer.onError, onDone: observer.onComplete);
+      _run = buildUseCaseTask();
```

Comment:
I don't understood why `buildUseCaseTask` should return a static reference of the usecase. Why not just implement the buildUseCaseTask and pass the reference to _run like `_run = buildUseCaseTask`. And then, in the isolate, execute it with the params passed through `BackgroundUseCaseParams`?